internal/api/token.go

```diff
@@ -577,6 +589,86 @@ func (a *API) IdTokenGrant(ctx context.Context, w http.ResponseWriter, r *http.R
 	return sendJSON(w, http.StatusOK, token)
 }
 
+func (a *API) OAuthPKCE(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+	db := a.db.WithContext(ctx)
+	var grantParams models.GrantParams
+
+	params := &PKCEGrantParams{}
+	body, err := getBodyBytes(r)
+	if err != nil {
+		return internalServerError("Could not read body").WithInternalError(err)
+	}
+
+	if err = json.Unmarshal(body, params); err != nil {
+		return badRequestError("invalid body: unable to parse JSON").WithInternalError(err)
+	}
+
+	if params.AuthCode == "" || params.CodeVerifier == "" {
+		return badRequestError("invalid request: both auth code and code verifier should be non-empty")
+	}
+
+	flowState, err := models.FindFlowStateByAuthCode(db, params.AuthCode)
+	// Sanity check in case user ID was not set properly
+	if models.IsNotFoundError(err) || flowState.UserID == uuid.Nil {
+		return forbiddenError("invalid oauth state, please ensure oauth redirect has successfully completed")
+	} else if err != nil {
+		return err
+	}
+
+	user, err := models.FindUserByID(db, flowState.UserID)
+	if err != nil {
+		return err
+	}
+
+	if flowState.AuthCode != params.AuthCode {
+		return forbiddenError("invalid auth code")
+	}
+
+	codeChallenge := flowState.CodeChallenge
+	codeChallengeAsString := string(codeChallenge[:])
+	codeVerifier := params.CodeVerifier
+
+	switch flowState.CodeChallengeMethod {
+	case models.SHA256.String():
+		encryptedCodeVerifier := sha256.Sum256([]byte(codeVerifier))
+		encodedCodeVerifier := base64.RawURLEncoding.EncodeToString(encryptedCodeVerifier[:])
```

Comment:
nit
```suggestion
		encodedCodeVerifier := base64.RawURLEncoding.EncodeToString(sha256.Sum256([]byte(codeVerifier))
```

Reply:
Oh I think this change will actually introduce an error though - see https://github.com/supabase/gotrue/pull/891#discussion_r1129200440 for details